core/src/main/java/org/elasticsearch/index/shard/ShardSplittingQuery.java

```diff
@@ -240,6 +294,46 @@ public float matchCost() {
             return 42; // that's obvious, right?
         }
     }
+
+    /**
+     * This TwoPhaseIterator marks all nested docs of matching parents as matches as well.
+     */
+    private final class NestedRoutingPartitionedDocIdSetIterator extends RoutingPartitionedDocIdSetIterator {
+        private final BitSet parentDocs;
+        private int nextParent = -1;
+        private boolean nextParentMatches;
+
+        NestedRoutingPartitionedDocIdSetIterator(LeafReader leafReader, Visitor visitor, BitSet parentDocs) {
+            super(leafReader, visitor);
+            this.parentDocs = parentDocs;
+        }
+
+        @Override
+        public boolean matches() throws IOException {
+            int doc = approximation.docID();
+            if (doc > nextParent) {
+                nextParent = parentDocs.nextSetBit(doc);
+                nextParentMatches = innerMatches(nextParent);
+            }
+            return nextParentMatches;
+        }
+    }
+
+    /*
+     * this is used internally to obtain a bitset for parent documents. We don't cache this since we never access the same reader more
+     * than once
+     */
```

Comment:
We warm this query per segment in `BitsetFilterCache#BitSetProducerWarmer` so why not using the per-segment cache directly ?

Reply:
we use this only as a delete by query which is executed on a `recovery-private` index writer. There is no point in cacheing it and it won't have a cache hit either.

Reply:
I left a comment

Reply:
Ok I missed the recovery-private thing. Thanks